blog/2024-11-effect-afterrendereffect/README.md

```diff
@@ -0,0 +1,504 @@
+  In this example no cleanup is required, but we wanted to mention the fact that long-running tasks (such as a timeout) should be cleaned up. The cleanup will be called before entering the same phase again, or if the effect itself is destroyed.
+  The `write` phase then passes the new height, `newHeight`, to the `read` phase.
+  Hint: Pass the same value to `read` (e.g. `return 100`) and you will see that the follow-up phase won't be executed.
+  Setting the same number twice won't be considered a change, so the `write` phase won't mark the `read` phase as dirty.
+
+3. **`read` Phase**: The final `read` phase logs the `newHeight`. 
+  We could also read from the DOM in that phase and store the result to a new signal. But in this example this work is not necessary, because the `earlyRead` is already doing that job.
+
+> We encourage you to scroll down to check out our Demo Application. 
+  Feel free to follow the hints in the comments to experiment with the specifics of each phase.
+
+
+## Migration Guide: From Angular's Lifecycle Hooks to Signal-Based Reactivity
+
+In April 2023, the Angular team outlined their vision of signal-based components in [RFC #49682](https://github.com/angular/angular/discussions/49682).
+The long-term goal is to phase out traditional lifecycle hooks, though the RFC discusses retaining `ngOnInit` and `ngOnDestroy`. (Now, we also have replacements for these.)
+The document proposed introducing `afterRenderEffect()` as part of a roadmap, and with Angular 19, the final vision of signal-based components is beginning to take shape.
+
+The addition of `effect()` and `afterRenderEffect()` showcases how Angular is moving in this direction. 
+These effects are more intuitive for managing component state changes and post-render interactions, thus making the old lifecycle hooks redundant.
+For instance, `afterRenderEffect()` is designed to handle tasks traditionally managed by `ngAfterViewInit` and `ngAfterViewChecked`.
+
+Migrating from Angular lifecycle hooks to `effect()` and `afterRenderEffect()` is straightforward:
+- **`ngOnInit` / `ngOnChanges`** → `effect()`: Handles signal-based logic and other state.
+- **`ngAfterViewInit` / `ngAfterViewChecked`** → `afterRenderEffect()`: Manages DOM manipulations post-render.
+
+Or to put it another way, here's a direct mapping:
+
+| Lifecycle Hook        | Replacement            |
+|-----------------------|------------------------|
+| `ngOnInit`            | `effect()`             |
+| `ngOnChanges`         | `effect()`             |
+| `ngAfterViewInit`     | `afterRenderEffect()`  |
+| `ngAfterViewChecked`  | `afterRenderEffect()`  |
+
+
+**Hint:** If you're transitioning away from classic lifecycle hooks, consider using [`DestroyRef`](https://angular.dev/api/core/DestroyRef).
+It allows you to set callbacks for cleanup or destruction tasks, so that you no longer need `ngOnDestroy` in your codebase. 
+
+
+## Reminder: `afterRenderEffect()` shouldn't be used in line-of-business code
```

Comment:
I don't know what you mean by "line-of-business code". Can we find another term?